test/framework/src/main/java/org/opensearch/test/TestLogsAppender.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.test;
+
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.appender.AbstractAppender;
+import org.apache.logging.log4j.core.config.Property;
+import org.apache.logging.log4j.core.layout.PatternLayout;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Test logs appender that provides functionality to extract specific logs/exception messages and wait for it to show up
+ * @opensearch.internal
+ */
+public class TestLogsAppender extends AbstractAppender {
+    private final List<String> capturedLogs = new ArrayList<>();
+    private final List<String> messagesToCapture;
+
+    public TestLogsAppender(List<String> messagesToCapture) {
+        super("TestAppender", null, PatternLayout.createDefaultLayout(), false, Property.EMPTY_ARRAY);
+        this.messagesToCapture = messagesToCapture;
+        start();
+    }
+
+    @Override
+    public void append(LogEvent event) {
+        if (shouldCaptureMessage(event.getMessage().getFormattedMessage())) capturedLogs.add(event.getMessage().getFormattedMessage());
+        if (event.getThrown() != null) {
+            if (shouldCaptureMessage(event.getThrown().toString())) capturedLogs.add(event.getThrown().toString());
+            for (StackTraceElement element : event.getThrown().getStackTrace())
+                if (shouldCaptureMessage(element.toString())) capturedLogs.add(element.toString());
+        }
+    }
```

Comment:
- formatting is done as per spotlessApply, not sure if i can change much here.
 - pattern match needs to be on message and any thrown exceptions because we also need to check exception message for our use-case

Reply:
- Got it. The single line if was bit long to read.
- Lets add strict asserts in the test to exactly match the expected part of the log message. It would be good to assert on the actual exception with the expected one (`event.getThrown`) if possible.